### ISSUE TITLE
[✨Feature Request]: add support aliases

### ISSUE THREAD

**Issue by PandaWorker** [fabian-hiller/valibot#163]
How about adding such functionality?

```js
import * as v from 'valibot';

const schema = v.object({
  name: v.alias('first_name', v.string())
});

const user = v.parse(schema, {
  first_name: 'anonym'
});

console.log(user.name) // anonym
```

```js
import * as v from 'valibot';

const schema = v.object({
  name: v.alias('user.name', v.string())
});

const user = v.parse(schema, {
 user: {
   name: 'anonym'
 }
});

console.log(user.name) // anonym
```

```js
import * as v from 'valibot';

const schema = v.object({
  name: v.alias('user[0]', v.string())
});

const user = v.parse(schema, {
 user: ['anonym', 'lastname']
});

console.log(user.name) // anonym
```

**Comment by fabian-hiller** [comment]
Thank you for sharing this idea. At the moment I am not sure if we should implement this. Is there a schema library that supports this feature? `alias` seems to be quite powerful, which will make it difficult to implement. Also, I am unsure how `alias` will interact with the other schema functions. I welcome feedback on this from others.

**Comment by fabian-hiller** [comment]
I am closing this issue due to inactivity. I welcome feedback if there is still interest in this feature.

**Comment by bdashore3** [comment]
Hi Fabian. I want to contribute back to this issue since aliases are features of validation libraries in other languages. However, I haven't seen this behavior used in TS validation libs which was pretty surprising.

Examples:
Python - [Pydantic](https://docs.pydantic.dev/latest/concepts/alias/#aliaspath-and-aliaschoices)
Rust - [Serde](https://serde.rs/field-attrs.html#alias)

Here's an example of a Pydantic model with aliases:
```python
class TestModel(BaseModel):
    static_field: Optional[float] = 0.0

    aliased_field: Optional[float] = Field(
        default=0.0,
        alias=AliasChoices("aliased_field", "dynamic_field")
    )
```

And from here, if `dynamic_field` is passed, it gets mapped to `aliased_field` and can be referenced in further code as such.

**Comment by fabian-hiller** [comment]
It is possible to implement it as a transformation action. Here is an example API:

```ts
import * as v from './dist/index.js';

const Schema = v.pipe(
  v.object({
    email: v.pipe(v.string(), v.email()),
    name: v.pipe(v.string(), v.nonEmpty()),
  }),
  v.aliasKey('name', 'first_name')
);

// Now input will be: { email: string, name: string }
// And output will be: { email: string, first_name: string }
```

Issue #981 is partly related.

**Comment by bdashore3** [comment]
Sorry, that's a bit hard to understand from my end. So, `aliasKey` is the custom transformation action? Do you mind pointing me to resources on how that can be created?

**Comment by fabian-hiller** [comment]
Yes, `aliasKey` would be the custom action. The name was just an initial idea.

Here is some context:
- We can't add the action to the key itself, because the transformation takes place in the outer object
- That's why the actions need to be placed in the pipeline of the object schema
- The action must transform not only the output data but also the output type
  - Something like `Omit<TObjectOutput, TOldKey> & { [TKey in TNewKey]: TObjectOutput[TOldKey] }` may work
- #981 is partly related – feel free to have a look
- I would probably copy and modify the [source code of `transform`](https://github.com/fabian-hiller/valibot/blob/main/library/src/actions/transform/transform.ts) when implementing the action

**Comment by bdashore3** [comment]
Thanks for the context. I'll take a look at it, but I don't believe I'm skilled enough in TS to properly PR this action. Can you consider reopening the issue and maybe someone else will take it up before I can?

**Comment by fabian-hiller** [comment]
Yes, don't worry. We could look into this when also working on #981.

**Comment by michachen** [comment]
```
import * as v from 'valibot';

const schema = v.object({
  name: v.alias('user.name', v.string())
});

const user = v.parse(schema, {
 user: {
   name: 'anonym'
 }
});
console.log(user.name) // anonym
```
@fabian-hiller @fabian-hiller 
The `user.name` alias string is like `lodash/get`
This is actually exactly what im looking for. Imagine you recieve a JSON from your server but the backend team wrote naming or even data nesting in a quirky  way. It would be nice to actually provide the original object and alias nested props from it. It will allow you to rearange your output object the way you feel like without making changes or rearranging namings/nesting on the backend side. And the the infered type would be also the way you like so its awesome IMO.

It would be possible to actually write the alias action if there was some way to pass the original object or some custom object via parse/safeParse config or global config.

**Comment by fabian-hiller** [comment]
I am not sure if `v.alias('user.firstName', v.string())` is technically possible without adding additional code to `object`. I think only `object` would be able to check if an `.alias` is defined to store the value under a different name. If this is the only option, I am not sure we should do this, as it would increase the bundle size for all users.

An API like this:

```ts
import * as v from 'valibot';

const Schema = v.pipe(
  v.object({
    email: v.pipe(v.string(), v.email()),
    name: v.pipe(v.string(), v.nonEmpty()),
  }),
  v.aliasKey('name', 'first_name')
);
```

would be easier to implement and would be more in line with our [mental model](https://valibot.dev/guides/mental-model/). However, it seems to be more complicated for the user as the alias cannot be defined directly by the key to be renamed.

Another option could be something like this (a combination of the first and second ideas):

```ts
const Schema = v.pipe(
  v.object({
    email: v.pipe(v.string(), v.email()),
    name: v.pipe(v.string(), v.nonEmpty(), v.alias('first_name')),
  }),
  v.aliasKeys()
);
```

In this case, `alias` would be a metadata action that just stores an info. `aliasKeys` would be a transformation action that reads the stored alias names and transforms the object accordingly.

**Comment by michachen** [comment]
@fabian-hiller The thing is that you don't only want to support aliasing key name to other key name, you would actually also want to alias (nice to have and easy to implement). What @PandaWorker and I suggest is aliasing a value/deeply nested value to some specific key name.
Those are two different things.

**Comment by mattrubin** [comment]
> … If this is the only option, I am not sure we should do this, as it would increase the bundle size for all users.

Would it be possible to implement this as a new function called something like `v.aliasedObject(…)` (or `v.objectWithAliasedKeys(…)`), so that users could easily opt in to the behavior, but it wouldn't impact the complexity of the existing `v.object(…)`? I'm envisioning something akin to your `v.pipe(v.object(…), v.aliasKeys())` example, but just wrapped up in a single function.
